stats.go

```diff
@@ -343,68 +341,27 @@ type statStore struct {
 
 	mu             sync.RWMutex
 	statGenerators []StatGenerator
-	stop           chan bool
-	wg             *sync.WaitGroup
+
+	wg sync.WaitGroup // for testing
 
 	sink Sink
 }
 
-func (s *statStore) Start(ticker *time.Ticker) {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
-	// if there is already a stop channel allocated, that means there
-	// is a ticker running - we will replace the ticker now.
-	if s.stop != nil {
-		s.stopLocked()
-	}
-
-	stopChan := make(chan bool, 1)
-	wg := &sync.WaitGroup{}
-	wg.Add(1)
-
-	s.stop = stopChan
-	s.wg = wg
-
-	go func() {
-		defer wg.Done()
-		for {
-			select {
-			case <-ticker.C:
-				s.Flush()
-			case <-stopChan:
-				return
-			}
+func (s *statStore) StartContext(ctx context.Context, ticker *time.Ticker) {
+	s.wg.Add(1)
+	defer s.wg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
```

Comment:
I think we want one last flush when we end
```suggestion
			s.Flush()
			return
```